adds feature specs

bjblock opened the pull request:
UPDATE
@raghubetina:
I have tests for pretty much everything now.  The outstanding questions are:

1. Should I move the validation tests to unit tests?  (See [my note below](#issuecomment-278116055))
2. Now that I've removed any use of the association methods in the tests, how should we test that the students added in the `has_many` and `belongs_to` per your [instruction in the readme](https://github.com/appdevfall16/msm_associations#association-helper-methods)
3. Should I remove all the tests for basic golden seven functionality?  (I made comments for those tests I think could be removed).

Once we've finalized everything, I'll push the code to the winter17 repo.

bjblock committed:
adds feature specs

raghubetina commented on spec/factories/actors.rb:
```diff
@@ -0,0 +1,4 @@
+FactoryGirl.define do
```
What do you think about putting all factories in a single `spec/factories.rb` file?

 - We don't want our factories to be overwritten by generators.
 - IIRC it is the style that thoughtbot encourages.

raghubetina commented on spec/factories/actors.rb:
```diff
@@ -0,0 +1,4 @@
+FactoryGirl.define do
+  factory :actor do
+  end
```
The philosophy that thoughtbot encourages for factories is to make them minimally viable objects -- no more and no less. In this case, it's doing less -- we should give it some default attributes so that they are useful. But no more than are required to pass validations.

In this particular assignment, we probably can't use the handy `association` method (since students may not have added the association helper methods yet), but I think we can do what we need with dynamic attributes.

Ref:

 - https://robots.thoughtbot.com/factories-should-be-the-bare-minimum
 - https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#associations
 - https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#dynamic-attributes

bjblock replied:
I struggled with this - what "default" attributes of an actor/movie/director/character would you add here?  All the attributes are specific to each record, so I'm not sure what default attributes would be useful.

raghubetina commented on spec/features/actors_feature_spec.rb:
```diff
@@ -0,0 +1,65 @@
+    create(:character, name: 'James Conway', movie: goodfellas, actor: bob)
+  end
+
+  context "index page" do
```
I like this structure of

 - describe
   - context
     - it

Makes `rspec --format documentation` read nicely 👍 

<img width="579" alt="screen shot 2017-02-07 at 12 34 18 pm" src="https://cloud.githubusercontent.com/assets/811217/22681149/c4fc5ae6-ed31-11e6-8ebb-5bc05895d7e8.png">

bjblock replied:
that's what i was going for 😄

raghubetina commented on spec/features/actors_feature_spec.rb:
```diff
@@ -0,0 +1,65 @@
+
+RSpec.describe "Actors", type: :feature do
+
+  before do
```
In general, I like to avoid `before` and `let`, to avoid [Mystery Guests]().

Ref:

 - https://robots.thoughtbot.com/mystery-guest
 - https://robots.thoughtbot.com/lets-not

bjblock replied:
I want to be sure there's enough data for accurate testing (i.e. multiple records of each) which, in this case, would seriously pollute the tests.  I recognize the students won't know how the data is created, but by using code in the tests to read the data (e.g. `Actor.all` or `Actor.last`), I figure the mystery guest problem is less of a factor.

raghubetina replied:
Okay, sounds good to me. Let's try it this way.

raghubetina commented on spec/features/actors_feature_spec.rb:
```diff
@@ -0,0 +1,65 @@
+
+  before do
+    scorsese = create(:director, name: "Martin Scorsese", dob: "November 17, 1942")
+    departed = create(:movie, title: "The Departed", year: 2006, duration: 151, director: scorsese)
```
Rather than hardcoding multiple movies, we might be better served by a sequence:

https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#sequences

bjblock replied:
I would typically agree, but if we want the tests to be more understandable to the students, real data seems more useful than sequenced data.

raghubetina replied:
Understood.

raghubetina commented on spec/features/actors_feature_spec.rb:
```diff
@@ -0,0 +1,65 @@
+    it "displays each actor's name and dob", points: 5 do
+      visit "/actors"
+
+      actors = Actor.all
```
What do you think of using `create_list` instead?

https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#building-or-creating-multiple-records

bjblock replied:
again here, going on the philosophy that we want the students to be able to read the tests, if the students go to inspect the test, I think they'll understand what `Actor.all` even if they don't fully understand where the actor data came from, whereas `create_list` hides a lot.  Plus, if we go that route, we definitely need to use sequencing to avoid uniqueness validation errors.

raghubetina commented on spec/features/charactors_feature_spec.rb:
```diff
@@ -0,0 +1,73 @@
+require "rails_helper"
```
There's a typo in this file name -- charactors => characters

raghubetina commented on spec/features/directors_feature_spec.rb:
```diff
@@ -0,0 +1,165 @@
+      end
+    end
+
+    it "reates a new movie after submitting the form", points: 10 do
```
Typo: "reates" => "creates"

raghubetina commented on spec/features/actors_feature_spec.rb:
```diff
@@ -0,0 +1,65 @@
+    end
+
+    it "displays a functional delete link for each actor", points: 5 do
+      visit "/actors"
```
In general, a lot of these tests are testing `starter_generators` Golden Seven functionality. Should we bother, outside of perhaps testing that they generated the right table/column names?

I am working on very detailed tests of Golden Seven in [this branch of Photogram G7](https://github.com/appdevwinter17/photogram_golden_seven/commits/solutions). We should be able to just copy-paste those over, or even just build them into the generator.

In this project, should we just concentrate on association- and validation-specific tasks?

bjblock replied:
Yeah, I wasn't sure if we want to test all the golden seven functionality (since that's what's in the target) or stick to the functionality specific to this assignment.  We can remove these tests if you'd like.

raghubetina commented on spec/features/charactors_feature_spec.rb:
```diff
@@ -0,0 +1,73 @@
+
+      characters = Character.all
+      characters.each do |character|
+        expect(page).to have_content(character.actor.name)
```
Will this work before students add the association helper methods? Should it?

bjblock replied:
the last instruction in the assignment is to add the `has_many` and `belongs_to` so I figured it was safe to test on those.  They're also possibly hints to the student that they should add them if they haven't already?  If those aren't requirements, then I can use `find_by`

raghubetina commented on spec/features/directors_feature_spec.rb:
```diff
@@ -0,0 +1,165 @@
+      directors.each do |director|
+        visit "/directors/#{director.id}"
+
+        expect(page).to have_content(director.movies.count)
```
Will this work before students add the association helper methods? Should it?

bjblock replied:
see my note above

raghubetina commented on spec/features/directors_feature_spec.rb:
```diff
@@ -0,0 +1,165 @@
+    it "displays a count of the director's movies", points: 5 do
+      directors = Director.all
+      directors.each do |director|
+        visit "/directors/#{director.id}"
```
In this case, what do you think about using `transient`?

https://github.com/thoughtbot/factory_girl/issues/527#issuecomment-18027737

bjblock replied:
not sure i understand

raghubetina replied:
That would have been in case we wanted to use Factories more rigorously, which it seems we've decided against. So nevermind :)

raghubetina commented on spec/features/directors_feature_spec.rb:
```diff
@@ -0,0 +1,165 @@
+      directors.each do |director|
+        visit "/directors/#{director.id}"
+
+        expect(page).to have_selector("form")
```
I think much of this testing logic can be copied from [Photogram G7](https://github.com/appdevwinter17/photogram_golden_seven/blob/solutions/spec/features/photos_spec.rb#L44)?

raghubetina commented on spec/features/directors_feature_spec.rb:
```diff
@@ -0,0 +1,165 @@
+    end
+
+    it "displays a hidden input field to associate a new movie to the director", points: 10 do
+      directors = Director.all
```
`create_list`?

bjblock replied:
see my note above about `create_list`

raghubetina commented on spec/features/directors_feature_spec.rb:
```diff
@@ -0,0 +1,165 @@
+        visit "/directors/#{director.id}"
+
+        expect(page).to have_selector("input[value='#{director.id}']", visible: false),
+          "expected to find a hidden input field with the director's id as the value"
```
Cool, I didn't know you could specify a message like this.

raghubetina commented on spec/features/directors_feature_spec.rb:
```diff
@@ -0,0 +1,165 @@
+      fill_in 'Dob', with: 'March 3, 1945'
+      click_button 'Create Director'
+
+      expect(Director.count).to eq(count_of_directors)
```
Can/should we make the description of the test match the expectation more closely? Something like `it "doesn't save the record if the name is blank"`?

bjblock replied:
good call, yes.  done.

raghubetina commented on spec/features/directors_feature_spec.rb:
```diff
@@ -0,0 +1,165 @@
+    end
+
+    it 'fails the validation if name is not unique' do
+      visit "/directors/new"
```
What do you think of unit tests for validations, something like

```ruby
first_director = create(:director)
duplicate_director = build(:director, name: first_director.name)
expect duplicate_director.to be_invalid
```
?

Philosophically, should we just use integration tests for everything so that we stay totally agnostic to implementation? If so, your approach is better.

bjblock replied:
I think in the past I've written unit tests and you encouraged us to only use integration tests, so was just following that methodology.  Unit tests would be more appropriate for these validation tests, but the benefit of solely using integration testing is that it will be easier for students to repeat the tests in their local apps to see why they're failing.

raghubetina commented on spec/features/movies_feature_spec.rb:
```diff
@@ -0,0 +1,124 @@
+
+      movies = Movie.all
+      movies.each do |movie|
+        expect(page).to have_content(movie.director.name)
```
Will this work before students add the association helper methods? Should it?

bjblock replied:
see my note above

bjblock replied:
instead of `expect(page).to have_content(movie.director.name)`,
would you prefer something like this?

```
director = Director.find_by(id: movie.director_id)
name = director.name
expect(page).to have_content(name)
```

raghubetina commented on spec/features/movies_feature_spec.rb:
```diff
@@ -0,0 +1,124 @@
+
+      directors = Director.all
+      directors.each do |director|
+        expect(page).to have_selector("select#director_id"), 'expected to find a select field for director_id in the form'
```
Is this looking for `id="director_id"` or `name="director_id"`? I think we need the latter?

raghubetina reviewed (commented):
Great start! In general,

 - Should we leave Golden Seven specs out? Our aim isn't to test `starter_generators`, after all.
 - We might want to consider using factories more idiomatically, and avoiding `before`.
 - Should we avoid assuming the student is using association helper methods?

bjblock committed:
moves factories to single file and updates some specs

bjblock commented:
I think one of my comments regarding unit tests is hidden now, so copying it here so we can discuss it:

> I think in the past I've written unit tests and you encouraged us to only use integration tests, so was just following that methodology. Unit tests would be more appropriate for these validation tests, but the benefit of solely using integration testing is that it will be easier for students to repeat the tests in their local apps to see why they're failing.

bjblock committed:
completes specs for readme instructions

bjblock committed:
removes use of association helper methods

bjblock commented:
in the latest commit, I removed all use of association methods.  So the tests will pass before they get to the last instruction to use `belongs_to` and `has_many`.  My question is, how should we test that they've added those?  I can duplicate all of those tests where I removed the association methods and repeat them with association methods (maybe in a different _context block_?).  Or do we just not test for that?

raghubetina commented:
> but the benefit of solely using integration testing is that it will be easier for students to repeat the tests in their local apps to see why they're failing.

I like this reasoning. Okay, let's see if we can stick with a purely "automating manually clicking" approach and see how that goes.

raghubetina commented:
> I can duplicate all of those tests where I removed the association methods and repeat them with association methods (maybe in a different context block?). Or do we just not test for that?

Hmm, to contradict [my previous comment](https://github.com/appdevfall16/msm_associations/pull/1#issuecomment-278198431), perhaps this really is a use case for unit tests; we could simply just check whether the object responds to particular methods.

I feel like this might still be consistent with "a purely 'automating manually clicking'" approach since we already have the implementation-agnostic integration spec, and this really is checking whether a particular implementation was used, and so a unit test is appropriate. What do you think?

bjblock committed:
renames factories so they don't contradict with generated factories

raghubetina commented on spec/factories.rb:
```diff
@@ -1,13 +1,13 @@
   end
 
-  factory :movie do
+  factory :new_movie, class: Movie do
```
It shouldn't be necessary to rename these, as long as they are not in `spec/factories/[MODEL NAME].rb`. Right? Or did you observe them being ignored?

bjblock committed:
adds config to avoid generating factories when using the model generator

bjblock committed:
adds config to disable any generation of specs;

also finalizes validation and association specs

bjblock commented:
Ok.  I think this is pretty good.
What do you think?

raghubetina commented:
Looks pretty good to me! Can you rebase onto the appdevwinter17 repo? I'll try running through on Windows.
